src/sagemaker_xgboost_container/algorithm_mode/callback.py

```diff
@@ -0,0 +1,205 @@
+import logging
+import os
+import tempfile
+import threading
+import queue
+from typing import Callable, Optional
+from xgboost import rabit
+from xgboost.core import CallbackEnv
+from xgboost.callback import _fmt_metric as fmt_metric
+
+
+TEMP_FILE_SUFFIX = ".sagemaker-ignore"
+FILE_LOCK_SUFFIX = ".sagemaker-uploading"
+FILE_SAFE_SUFFIX = ".sagemaker-uploaded"
+
+CHECKPOINT_FILENAME = "xgboost-checkpoint"
+CHECKPOINT_NUM_DIGITS = 12
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+def _zero_pad(iteration: int) -> str:
+
+    iter_num = format(iteration, "0{}".format(CHECKPOINT_NUM_DIGITS))
+
+    return iter_num
+
+
+# modified from https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/callback.py
+def print_evaluation(period=1, show_stdv=True, start_iteration=0):
+    """Create a callback that print evaluation result.
+
+    We print the evaluation results every **period** iterations
+    and on the first and the last iterations.
+
+    Parameters
+    ----------
+    period : int
+        The period to log the evaluation results
+
+    show_stdv : bool, optional
+         Whether show stdv if provided
+
+    Returns
+    -------
+    callback : function
+        A callback that print evaluation every period iterations.
+    """
+    def callback(env):
+        """internal function"""
+        if env.rank != 0 or (not env.evaluation_result_list) or period is False or period == 0:
+            return
+        i = env.iteration
+        if i % period == 0 or i + 1 == env.begin_iteration or i + 1 == env.end_iteration:
+            msg = '\t'.join([fmt_metric(x, show_stdv) for x in env.evaluation_result_list])
+            rabit.tracker_print('[%d]\t%s\n' % (i + start_iteration, msg))
+    return callback
+
+
+class SaveCheckpoint:
```

Comment:
Overall this class seems much more complex than needed, or not complex enough. If we are deleting files based on 'flag' files why do we need to maintain a queue to remove files; wouldn't it be easier to just remove all files that we recognize as done? 

On the other side, I'm not as familiar with spot instances; what exactly is the behavior when a spot instance is taken away? Does the kernel recognize it as a interrupt? Is this process supposed to block on uploading the files, or do we just take it as best attempt and take whatever was managed to be saved in s3?